org.librarysimplified.r2.views/src/main/java/org/librarysimplified/r2/views/SR2ReaderFragment.kt

```diff
@@ -101,57 +107,70 @@ class SR2ReaderFragment : Fragment() {
     FluentFuture.from(controllerFuture)
       .transform(
         Function<SR2ControllerType, Unit> { controller ->
-          this.onBookOpenSucceeded(controller!!)
+          onBookOpenSucceeded(controller!!)
         },
         MoreExecutors.directExecutor()
       )
       .catching(
         Throwable::class.java,
-        Function<Throwable, Unit> { exception ->
-          this.onBookOpenFailed(exception!!)
+        Function<Throwable, Unit> { e ->
+          onBookOpenFailed(e!!)
         }, MoreExecutors.directExecutor()
       )
   }
 
-  private fun onBookOpenSucceeded(controller: SR2ControllerType) {
-    this.logger.debug("onBookOpenSucceeded")
+  override fun onStop() {
+    super.onStop()
+    controllerSubscription?.dispose()
+    controller?.viewDisconnect()
+  }
 
-    this.runOnUIThread {
-      this.onBookOpenSucceededUI(controller)
-    }
+  private fun onBookOpenSucceeded(controller: SR2ControllerType) {
+    logger.debug("onBookOpenSucceeded")
+    runOnUIThread { onBookOpenSucceededUI(controller) }
   }
 
   @UiThread
   private fun onBookOpenSucceededUI(controller: SR2ControllerType) {
     this.controller = controller
-    controller.viewConnect(this.webView)
-    this.controllerHost.onControllerBecameAvailable(controller)
+    controller.viewConnect(webView)
+
+    controllerSubscription =
+      controller.events.subscribe { onControllerEvent(it) }
+
+    controllerHost.onControllerBecameAvailable(controller)
   }
 
-  private fun onBookOpenFailed(exception: Throwable) {
-    this.logger.error("onBookOpenFailed: ", exception)
+  private fun onBookOpenFailed(e: Throwable) {
+    logger.error("onBookOpenFailed: ", e)
+    runOnUIThread { onBookOpenFailedUI(e) }
+  }
 
-    this.runOnUIThread {
-      this.onBookOpenFailedUI(exception)
-    }
+  @UiThread
+  private fun onBookOpenFailedUI(e: Throwable) {
+    TODO()
```

Comment:
Not sure what's going on here.

Reply:
I think this method was just empty before. The only change I made was add the "TODO()" but github is struggling with the diff.

Reply:
Right!